drivers/pci/controller/vmd.c

```diff
@@ -767,17 +843,226 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+		.start = vmd->busn_start[VMD_ROOTBUS0],
+		.end = vmd->busn_start[VMD_ROOTBUS0] +
+		       (resource_size(res) >> 20) - 1,
+		.flags = IORESOURCE_BUS | IORESOURCE_PCI_FIXED,
+	};
+
+	if (vmd_has_pch_rootbus(vmd)) {
+		pci_read_config_word(vmd->dev, PCI_REG_BUSRANGE0,
+				     &ioc_bus_range);
+		pci_read_config_word(vmd->dev, PCI_REG_BUSRANGE1,
+				     &pch_bus_range);
+
+		/*
+		 * Resize CPU IOC CFGBAR range to make space for PCH owned
+		 * devices by adjusting range end with value stored in
+		 * PCI_REG_BUSRANGE0 register.
+		 */
+		vmd->resources[VMD_RESOURCE_CFGBAR].start =
+			VMD_ROOTBUS_RANGE_START(ioc_bus_range);
+		vmd->resources[VMD_RESOURCE_CFGBAR].end =
+			VMD_ROOTBUS_RANGE_END(ioc_bus_range);
+
+		vmd->resources[VMD_RESOURCE_PCH_CFGBAR] = (struct resource){
+			.name = "VMD CFGBAR PCH",
+			.start = VMD_ROOTBUS_RANGE_START(pch_bus_range),
+			.end = VMD_ROOTBUS_RANGE_END(pch_bus_range),
+			.flags = IORESOURCE_BUS | IORESOURCE_PCI_FIXED,
+			.parent = &vmd->resources[VMD_RESOURCE_CFGBAR],
+		};
+	}
+}
+
+static void vmd_configure_membar(struct vmd_dev *vmd,
+				 u8 membar_number, u8 resource_number,
+				 resource_size_t start_offset,
+				 resource_size_t end_offset,
+				 struct resource *parent)
 {
-	struct pci_sysdata *sd = &vmd->sysdata;
 	struct resource *res;
 	u32 upper_bits;
 	unsigned long flags;
+	char name[16];
+
+	res = &vmd->dev->resource[membar_number];
+	upper_bits = upper_32_bits(res->end);
+	flags = res->flags & ~IORESOURCE_SIZEALIGN;
+	if (!upper_bits)
+		flags &= ~IORESOURCE_MEM_64;
+
+	snprintf(name, sizeof(name), "VMD MEMBAR%d", membar_number/2);
+
+	if (!parent)
+		parent = res;
+
+	if (resource_number > VMD_RESOURCE_MEMBAR2)
+		strncat(name, " PCH", 4);
+
+	vmd->resources[resource_number] = (struct resource){
+		.name = name,
+		.start = res->start + start_offset,
+		.end = res->end - end_offset,
+		.flags = flags,
+		.parent = parent,
+	};
+}
+
+static void vmd_configure_membar1_membar2(struct vmd_dev *vmd,
+					  resource_size_t membar2_offset)
+{
+	vmd_configure_membar(vmd, VMD_MEMBAR1, VMD_RESOURCE_MEMBAR1, 0, 0,
+			     NULL);
+	vmd_configure_membar(vmd, VMD_MEMBAR2, VMD_RESOURCE_MEMBAR2,
+			     membar2_offset, 0, NULL);
+
+	if (vmd_has_pch_rootbus(vmd)) {
+		u32 pch_membar1_offset = 0;
+		u64 pch_membar2_offset = 0;
+		u32 reg;
+
+		pci_read_config_dword(vmd->dev, PCI_MEMBAR1_OFFSET,
+				      &pch_membar1_offset);
+
+		pci_read_config_dword(vmd->dev, PCI_MEMBAR2_OFFSET1, &reg);
+		pch_membar2_offset = reg;
+
+		pci_read_config_dword(vmd->dev, PCI_MEMBAR2_OFFSET2, &reg);
+		pch_membar2_offset |= (u64)reg << 32;
+
+		/*
+		 * Resize CPU IOC MEMBAR1 and MEMBAR2 ranges to make space
+		 * for PCH owned devices by adjusting range end with values
+		 * stored in PCI_MEMBAR1_OFFSET and PCI_MEMBAR2_OFFSET registers
+		 */
+		vmd_configure_membar(vmd, VMD_MEMBAR1, VMD_RESOURCE_MEMBAR1, 0,
+				     pch_membar1_offset, NULL);
+		vmd_configure_membar(vmd, VMD_MEMBAR2, VMD_RESOURCE_MEMBAR2,
```

Comment:
What I don' like here is repeated initialization please consider reverting order. or use if/else

Reply:
Fixed.